### PR TITLE
Box decode assign op

### PR DESCRIPTION
**Pull request by dyning**
add Box decode assign op for cascade RCNN

### CODE REVIEW

File: paddle/fluid/operators/detection/box_decoder_and_assign_op.cc [hunk 1]
```diff
@@ -0,0 +1,148 @@
+/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/box_decoder_and_assign_op.cc]: 2018 -> 2019, same as below

File: paddle/fluid/operators/detection/box_decoder_and_assign_op.cc [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+}  // namespace operators
+}  // namespace paddle
+
+namespace ops = paddle::operators;
+REGISTER_OPERATOR(box_decoder_and_assign, ops::BoxDecoderAndAssignOp,
+                  ops::BoxDecoderAndAssignOpMaker,
+                  paddle::framework::EmptyGradOpMaker);
+REGISTER_OP_CPU_KERNEL(
+    box_decoder_and_assign,
+    ops::BoxDecoderAndAssignKernel<paddle::platform::CPUDeviceContext, float>);
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/box_decoder_and_assign_op.cc]: add double type

File: python/paddle/fluid/layers/detection.py [hunk 3]
```diff
@@ -2221,3 +2222,40 @@ class number
     output.stop_gradient = True
 
     return output
+
+
+@templatedoc()
+def box_decoder_and_assign(prior_box, prior_box_var, target_box, box_score,
```
Comment (jerrywgz) [inline python/paddle/fluid/layers/detection.py:2228]: It's better to give an example

File: paddle/fluid/operators/detection/box_decoder_and_assign_op.h [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    auto* prior_box_data = prior_box->data<T>();
+    auto* prior_box_var_data = prior_box_var->data<T>();
+    auto* box_score_data = box_score->data<T>();
+    output_box->mutable_data<T>({roi_num, class_num * 4}, context.GetPlace());
+    output_assign_box->mutable_data<T>({roi_num, 4}, context.GetPlace());
+    T* output_box_data = output_box->data<T>();
+    T* output_assign_box_data = output_assign_box->data<T>();
+    const float bbox_clip = static_cast<float>(context.Attr<float>("box_clip"));
+
+    std::vector<float> tmp_score(class_num);
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/box_decoder_and_assign_op.h]: what's this for?

File: paddle/fluid/operators/detection/box_decoder_and_assign_op.cc [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        "bottom coordinate of the box. This tensor can contain LoD "
+        "information to represent a batch of inputs. One instance of this "
+        "batch can contain different numbers of entities.");
+    AddInput(
+        "BoxScore",
+        "(LoDTensor or Tensor) This input can be a 2-D LoDTensor with shape "
+        "[N, classnum], each box is represented as [classnum] which is "
+        "the classification probabilities.");
+    AddAttr<float>("box_clip",
+                   "(float, default 4.135, np.log(1000. / 16.)) "
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/box_decoder_and_assign_op.cc:115]: For generate proposals op, we set box_clip as a const value inside op. Is this attr necessary to expose to users?

File: paddle/fluid/operators/detection/box_decoder_and_assign_op.h [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        if (score > max_score && j > 0) {
+          max_score = score;
+          max_j = j;
+        }
+      }
+
+      if (max_j > 0) {
+        for (int pno = 0; pno < 4; pno++) {
+          output_assign_box_data[i * 4 + pno] =
+              output_box_data[i * class_num * 4 + max_j * 4 + pno];
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/box_decoder_and_assign_op.h:91]: It's better to use memory::Copy

File: paddle/fluid/operators/detection/box_decoder_and_assign_op.cu [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    if (max_j > 0) {
+      for (int pno = 0; pno < 4; pno++) {
+        output_assign_box_data[i * 4 + pno] =
+            output_box_data[i * class_num * 4 + max_j * 4 + pno];
+      }
+    } else {
+      for (int pno = 0; pno < 4; pno++) {
+        output_assign_box_data[i * 4 + pno] = prior_box_data[i * 4 + pno];
+      }
+    }
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/box_decoder_and_assign_op.cu:90]: It's better to use memory Copy

File: paddle/fluid/operators/detection/box_decoder_and_assign_op.cu [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+  }
+};
+
+}  // namespace operators
+}  // namespace paddle
+
+namespace ops = paddle::operators;
+REGISTER_OP_CUDA_KERNEL(box_decoder_and_assign,
+                        ops::BoxDecoderAndAssignCUDAKernel<
+                            paddle::platform::CUDADeviceContext, float>);
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/box_decoder_and_assign_op.cu]: add double type

File: paddle/fluid/operators/detection/box_decoder_and_assign_op.cc (hunk 1 above)
Comment (dyning) [reply paddle/fluid/operators/detection/box_decoder_and_assign_op.cc]: thanks

File: paddle/fluid/operators/detection/box_decoder_and_assign_op.cc (hunk 5 above)
Comment (dyning) [reply paddle/fluid/operators/detection/box_decoder_and_assign_op.cc:115]: for better using

File: paddle/fluid/operators/detection/box_decoder_and_assign_op.h (hunk 4 above)
Comment (dyning) [reply paddle/fluid/operators/detection/box_decoder_and_assign_op.h]: clipping the box extended the image

**Comment by dyning** [comment]
to https://github.com/PaddlePaddle/Paddle/pull/16048

**Pull request closed without merging**